### PR TITLE
BaSP:TG-17:Added seeds and test for Projects

### PR DESCRIPTION
**Pull request by SteampunkMatu**

### CODE REVIEW

File: src/models/Projects.js [hunk 1]
```diff
@@ -7,14 +7,10 @@ const projectSchema = new Schema({
   startDate: { type: Date, required: true },
   endDate: { type: Date, required: true },
   employees: [{
+    _id: false, // Para evitar que mongoose autogenere un nuevo id
```
Comment (lmalvasia) [inline src/models/Projects.js:10]: Remove the comment here

File: src/seed/projects.js [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+  endDate: '2022-03-25T03:00:00.000Z',
+  employees: [
+    {
+      _id: mongoose.Types.ObjectId('634c68f3658f142935ea7f6e'),
+      rate: 20,
+      role: 'DEV',
+    },
+  ],
+},
+];
```
Comment (lmalvasia) [inline src/seed/projects.js:63]: Move this at the end of the line 62.

File: src/test/projects.test.js [hunk 3]
```diff
@@ -1,3 +1,146 @@
...
+        role: 'DEV',
+      },
+    ],
+  };
+  beforeAll(async () => {
+    await projects.collection.insertMany(projectSeed);
+  });
+
+  describe('GET /projects', () => {
+    test('should return the status code 200 and the list of projects', async () => {
```
Comment (lmalvasia) [inline src/test/projects.test.js:29]: All of those tests could be checked inside only one test.
You can have more than one `expect` inside a test

File: src/test/projects.test.js [hunk 4]
```diff
@@ -1,3 +1,146 @@
...
+      expect(response.body.error).toBeFalsy();
+      expect(response.body.data).toMatchObject([{
+        name: mockedProject.name,
+        clientName: mockedProject.clientName,
+        description: mockedProject.description,
+        startDate: mockedProject.startDate,
+        endDate: mockedProject.endDate,
+        employees: mockedProject.employees,
+      }]);
+    });
```
Comment (lmalvasia) [inline src/test/projects.test.js:57]: Add a new line between each test.

File: src/test/projects.test.js [hunk 5]
```diff
@@ -1,3 +1,146 @@
...
+    test('should not create a project', async () => {
```
Comment (lmalvasia) [inline src/test/projects.test.js:58]: why?
```suggestion
    test('should not create a project when there is not body', async () => {
```

File: src/test/projects.test.js [hunk 6]
```diff
@@ -1,3 +1,146 @@
...
+      const response = await request(app).post('/projects').send();
+      expect(response.status).toBe(400);
+      expect(response.body.error).toBeTruthy();
+      expect(response.body.data).toBeUndefined();
+    });
+    test('Project not created because incorrect Project name', async () => {
+      const response = await request(app).post('/projects').send({ mockedProject, name: 'a' });
+
+      expect(response.status).toBe(400);
+      expect(response.body.error).toBeTruthy();
```
Comment (lmalvasia) [inline src/test/projects.test.js:68]: Could you add a new expect to check the message?
expect(response.body.message).toEqual('....');

File: src/test/projects.test.js [hunk 7]
```diff
@@ -1,3 +1,146 @@
...
+        employees: mockedProject.employees,
+      }]);
+    });
+    test('should not create a project', async () => {
+      const response = await request(app).post('/projects').send();
+      expect(response.status).toBe(400);
+      expect(response.body.error).toBeTruthy();
+      expect(response.body.data).toBeUndefined();
+    });
+    test('Project not created because incorrect Project name', async () => {
```
Comment (lmalvasia) [inline src/test/projects.test.js:64]:
```suggestion
    test('should not create a project when the user rend an invalid project name', async () => {
```

File: src/test/projects.test.js [hunk 8]
```diff
@@ -1,3 +1,146 @@
...
+      const response = await request(app).get('/projects').send();
+      expect(response.body.error).toBeFalsy();
+    });
+    test('return more than one project', async () => {
+      const response = await request(app).get('/projects').send();
+      expect(response.body.data.length).toEqual(4);
+    });
+  });
+
+  describe('POST /projects', () => {
```
Comment (lmalvasia) [inline src/test/projects.test.js:43]: Check all the descriptions and you have to check the error message (see comments for the first test).

File: src/test/projects.test.js [hunk 9]
```diff
@@ -1,3 +1,146 @@
...
+    });
+    test('Project not created because incorrect Project employee', async () => {
+      const response = await request(app).post('/projects').send(({ mockedProject, employee: 'a' }));
+      expect(response.status).toBe(400);
+      expect(response.body.error).toBeTruthy();
+    });
+  });
+
+  describe('PUT /projects', () => {
+    beforeEach(async () => {
```
Comment (lmalvasia) [inline src/test/projects.test.js:98]: Remove this beforeEach. It's not necessary

File: src/test/projects.test.js [hunk 10]
```diff
@@ -1,3 +1,146 @@
...
+      newProject = await projects.create(mockedProject);
+    });
+    afterEach(async () => {
+      await projects.findByIdAndUpdate(newProject.id);
```
Comment (lmalvasia) [inline src/test/projects.test.js:102]: Remove this.

File: src/test/projects.test.js [hunk 11]
```diff
@@ -1,3 +1,146 @@
...
+    });
+    test('should edit a project', async () => {
+      const response = await request(app).put(`/projects/${newProject.id}`).send({ ...mockedProject, clientName: 'newName' });
+      expect(response.status).toBe(200);
+      expect(response.body.error).toBeFalsy();
```
Comment (lmalvasia) [inline src/test/projects.test.js:107]: You have to check the response.body.data

File: src/test/projects.test.js [hunk 12]
```diff
@@ -1,3 +1,146 @@
...
+    });
+    test('should not edit the project because an error', async () => {
```
Comment (lmalvasia) [inline src/test/projects.test.js:109]:
```suggestion
    test('should not edit a project when the user sends a invalid project id', async () => {
```

File: src/test/projects.test.js [hunk 13]
```diff
@@ -1,3 +1,146 @@
...
+      const response = await request(app).put(`/projects/${invalidProjectId}`).send();
+      expect(response.status).toBe(400);
+      expect(response.body.error).toBeTruthy();
+      expect(response.body.data).toBeUndefined();
+    });
+  });
+
+  describe('DELETE /projects', () => {
+    beforeEach(async () => {
```
Comment (lmalvasia) [inline src/test/projects.test.js:118]: Remove this.

File: src/test/projects.test.js [hunk 14]
```diff
@@ -1,3 +1,146 @@
...
+      newProject = await projects.create(mockedProject);
+    });
+    test('should remove the project', async () => {
+      const response = await request(app).delete(`/projects/${newProject.id}`).send();
+      expect(response.status).toBe(200);
+      expect(response.body.error).toBeFalsy();
+    });
```
Comment (lmalvasia) [inline src/test/projects.test.js:125]: You can add a test sending an invalid id.

File: src/test/projects.test.js [hunk 15]
```diff
@@ -1,3 +1,146 @@
...
+  });
+
+  describe('GetById /projects', () => {
+    beforeEach(async () => {
```
Comment (lmalvasia) [inline src/test/projects.test.js:129]: remove this.

File: src/test/projects.test.js [hunk 16]
```diff
@@ -1,3 +1,146 @@
...
+      newProject = await projects.create(mockedProject);
+    });
+    test('should get a project by ID', async () => {
+      const response = await request(app).get(`/projects/${newProject.id}`).send();
+      expect(response.status).toBe(200);
+      expect(response.body.data).toBeDefined();
+      const projectFound = await projects.findById(newProject.id);
```
Comment (lmalvasia) [inline src/test/projects.test.js:136]: remove this.

File: src/test/projects.test.js [hunk 17]
```diff
@@ -1,3 +1,146 @@
...
+  });
+
+  describe('GetById /projects', () => {
+    beforeEach(async () => {
+      newProject = await projects.create(mockedProject);
+    });
+    test('should get a project by ID', async () => {
+      const response = await request(app).get(`/projects/${newProject.id}`).send();
+      expect(response.status).toBe(200);
+      expect(response.body.data).toBeDefined();
```
Comment (lmalvasia) [inline src/test/projects.test.js:135]: this test is invalid. You are getting an undefined data and you should get the project.

File: src/test/projects.test.js [hunk 18]
```diff
@@ -1,3 +1,146 @@
...
+      expect(response.body.error).toBeFalsy();
+    });
+  });
+
+  describe('GetById /projects', () => {
+    beforeEach(async () => {
+      newProject = await projects.create(mockedProject);
+    });
+    test('should get a project by ID', async () => {
+      const response = await request(app).get(`/projects/${newProject.id}`).send();
```
Comment (lmalvasia) [inline src/test/projects.test.js:133]: You have to use the id of one of the projects that you have inside the seeds.

File: src/test/projects.test.js [hunk 19]
```diff
@@ -1,3 +1,146 @@
...
+      expect(response.status).toBe(200);
+      expect(response.body.data).toBeDefined();
+      const projectFound = await projects.findById(newProject.id);
+      expect(projectFound).toBeDefined();
+    });
+    test('Project ID not found', async () => {
+      const response = await request(app).get(`/projects/${invalidProjectId}`).send();
+      expect(response.status).toBe(404);
+      expect(response.body.error).toBeTruthy();
```
Comment (lmalvasia) [inline src/test/projects.test.js:142]: Can you add an expect to check the error message?

File: src/models/Projects.js (hunk 1 above)
Comment (SteampunkMatu) [reply src/models/Projects.js:10]: Done

File: src/seed/projects.js (hunk 2 above)
Comment (SteampunkMatu) [reply src/seed/projects.js:63]: Done

File: src/test/projects.test.js [hunk 20]
```diff
@@ -1,3 +1,146 @@
-test('', () => {
-  expect(2).toBe(2);
+import request from 'supertest';
+import mongoose from 'mongoose';
+import app from '../app';
+import projects from '../models/Projects';
+import projectSeed from '../seed/projects';
+
+describe('Projects -Test', () => {
```
[the remark this answers is not in the document]
Comment (SteampunkMatu) [reply src/test/projects.test.js:7]: Fixed

File: src/test/projects.test.js (hunk 3 above)
Comment (SteampunkMatu) [reply src/test/projects.test.js:29]: Okey, i'll fix that

File: src/test/projects.test.js (hunk 4 above)
Comment (SteampunkMatu) [reply src/test/projects.test.js:57]: Done

File: src/test/projects.test.js (hunk 5 above)
Comment (SteampunkMatu) [reply src/test/projects.test.js:58]: Done

File: src/test/projects.test.js (hunk 7 above)
Comment (SteampunkMatu) [reply src/test/projects.test.js:64]: Done

File: src/test/projects.test.js (hunk 9 above)
Comment (SteampunkMatu) [reply src/test/projects.test.js:98]: Done

File: src/test/projects.test.js (hunk 10 above)
Comment (SteampunkMatu) [reply src/test/projects.test.js:102]: Done

File: src/test/projects.test.js (hunk 11 above)
Comment (SteampunkMatu) [reply src/test/projects.test.js:107]: Done

File: src/test/projects.test.js (hunk 12 above)
Comment (SteampunkMatu) [reply src/test/projects.test.js:109]: done

File: src/test/projects.test.js (hunk 13 above)
Comment (SteampunkMatu) [reply src/test/projects.test.js:118]: Done

File: src/test/projects.test.js (hunk 14 above)
Comment (SteampunkMatu) [reply src/test/projects.test.js:125]: Done

File: src/test/projects.test.js (hunk 15 above)
Comment (SteampunkMatu) [reply src/test/projects.test.js:129]: Done

File: src/test/projects.test.js (hunk 19 above)
Comment (SteampunkMatu) [reply src/test/projects.test.js:142]: Done

File: src/test/projects.test.js (hunk 8 above)
Comment (SteampunkMatu) [reply src/test/projects.test.js:43]: Done

File: src/test/projects.test.js (hunk 6 above)
Comment (SteampunkMatu) [reply src/test/projects.test.js:68]: Done

File: src/test/projects.test.js (hunk 16 above)
Comment (SteampunkMatu) [reply src/test/projects.test.js]: Done

File: src/test/projects.test.js (hunk 17 above)
Comment (SteampunkMatu) [reply src/test/projects.test.js]: Done

File: src/test/projects.test.js (hunk 18 above)
Comment (SteampunkMatu) [reply src/test/projects.test.js]: Done

File: src/test/projects.test.js [hunk 21]
```diff
@@ -1,3 +1,145 @@
...
+    });
+
+    test('should not create a project when there is not body', async () => {
+      const response = await request(app).post('/projects').send();
+      expect(response.status).toBe(400);
+      expect(response.body.error).toBeTruthy();
+      expect(response.body.data).toBeUndefined();
+    });
+
+    test('should not create a project when the user rend an invalid project name', async () => {
```
Comment (lmalvasia) [inline src/test/projects.test.js:59]:
```suggestion
    test('should not create a project when the user sends an invalid project name', async () => {
```

File: src/test/projects.test.js [hunk 22]
```diff
@@ -1,3 +1,145 @@
...
+      const response = await request(app).post('/projects').send({ mockedProject, name: 'a' });
+      expect(response.status).toBe(400);
+      expect(response.body.error).toBeTruthy();
+      expect(response.body.message).toEqual('There was an error: "name" length must be at least 3 characters long');
+    });
+
+    test('Project not created because incorrect Project description', async () => {
```
Comment (lmalvasia) [inline src/test/projects.test.js:66]:
```suggestion
    test('should not create a project when the user sends an invalid project description', async () => {
```

File: src/test/projects.test.js [hunk 23]
```diff
@@ -1,3 +1,145 @@
...
+      const response = await request(app).post('/projects').send(({ mockedProject, description: 'a' }));
+      expect(response.status).toBe(400);
+      expect(response.body.error).toBeTruthy();
+      expect(response.body.message).toEqual('There was an error: "name" is required');
+    });
+
+    test('Project not created because incorrect Project end date', async () => {
```
Comment (lmalvasia) [inline src/test/projects.test.js:73]:
```suggestion
    test('should not create a project when the user sends an invalid project endDate', async () => {
```

File: src/test/projects.test.js [hunk 24]
```diff
@@ -1,3 +1,145 @@
...
+      const response = await request(app).post('/projects').send(({ mockedProject, endDate: 'a' }));
+      expect(response.status).toBe(400);
+      expect(response.body.error).toBeTruthy();
+      expect(response.body.message).toEqual('There was an error: "name" is required');
+    });
+
+    test('Project not created because incorrect Client name', async () => {
```
Comment (lmalvasia) [inline src/test/projects.test.js:80]:
```suggestion
    test('should not create a project when the user sends an invalid project clientName', async () => {
```

File: src/test/projects.test.js [hunk 25]
```diff
@@ -1,3 +1,145 @@
...
+      const response = await request(app).post('/projects').send(({ mockedProject, clientName: 'a' }));
+      expect(response.status).toBe(400);
+      expect(response.body.error).toBeTruthy();
+      expect(response.body.message).toEqual('There was an error: "name" is required');
+    });
+
+    test('Project not created because incorrect Project start date', async () => {
```
Comment (lmalvasia) [inline src/test/projects.test.js:87]:
```suggestion
    test('should not create a project when the user sends an invalid project startDate', async () => {
```

File: src/test/projects.test.js [hunk 26]
```diff
@@ -1,3 +1,145 @@
...
+      const response = await request(app).post('/projects').send(({ mockedProject, startDate: 'a' }));
+      expect(response.status).toBe(400);
+      expect(response.body.error).toBeTruthy();
+      expect(response.body.message).toEqual('There was an error: "name" is required');
+    });
+
+    test('Project not created because incorrect Project employee', async () => {
```
Comment (lmalvasia) [inline src/test/projects.test.js:94]:
```suggestion
    test('should not create a project when the user sends an invalid project employee', async () => {
```

File: src/test/projects.test.js [hunk 27]
```diff
@@ -1,3 +1,145 @@
...
+    });
+  });
+
+  describe('PUT /projects', () => {
+    test('should edit a project', async () => {
+      // eslint-disable-next-line no-underscore-dangle
+      const response = await request(app).put(`/projects/${projectSeed[0]._id}`).send({ ...mockedProject, clientName: 'newName' });
+      expect(response.status).toBe(200);
+      expect(response.body.data).toBeDefined();
+    });
```
Comment (lmalvasia) [inline src/test/projects.test.js:108]: Missing new line

File: src/test/projects.test.js [hunk 28]
```diff
@@ -1,3 +1,145 @@
...
+  });
+
+  describe('DELETE /projects', () => {
+    test('should remove the project', async () => {
+      // eslint-disable-next-line no-underscore-dangle
+      const response = await request(app).delete(`/projects/${projectSeed[1]._id}`).send();
+      expect(response.status).toBe(200);
+      expect(response.body.error).toBeFalsy();
+    });
```
Comment (lmalvasia) [inline src/test/projects.test.js:123]: Missing new line

File: src/test/projects.test.js [hunk 29]
```diff
@@ -1,3 +1,145 @@
...
+      expect(response.body.error).toBeTruthy();
+    });
+  });
+
+  describe('GetById /projects', () => {
+    test('should get a project by ID', async () => {
+      // eslint-disable-next-line no-underscore-dangle
+      const response = await request(app).get(`/projects/${projectSeed[2]._id}`).send();
+      expect(response.status).toBe(200);
+    });
```
Comment (lmalvasia) [inline src/test/projects.test.js:136]: Missing new line

File: src/test/projects.test.js [hunk 30]
```diff
@@ -1,3 +1,145 @@
...
+  });
+
+  describe('DELETE /projects', () => {
+    test('should remove the project', async () => {
+      // eslint-disable-next-line no-underscore-dangle
+      const response = await request(app).delete(`/projects/${projectSeed[1]._id}`).send();
+      expect(response.status).toBe(200);
+      expect(response.body.error).toBeFalsy();
+    });
+    test('Cant remove the project, invalid ID', async () => {
```
Comment (lmalvasia) [inline src/test/projects.test.js:124]:
```suggestion
    test('should return an error when the user sends an invalid id, async () => {
```

File: src/test/projects.test.js [hunk 31]
```diff
@@ -1,3 +1,145 @@
...
+      const response = await request(app).delete(`/projects/${invalidProjectId}`).send();
+      expect(response.status).toBe(404);
+      expect(response.body.error).toBeTruthy();
```
Comment (lmalvasia) [inline src/test/projects.test.js:127]: Check the error message

File: src/test/projects.test.js (hunk 21 above)
Comment (SteampunkMatu) [reply src/test/projects.test.js:59]: Done

File: src/test/projects.test.js (hunk 22 above)
Comment (SteampunkMatu) [reply src/test/projects.test.js:66]: Done

File: src/test/projects.test.js (hunk 23 above)
Comment (SteampunkMatu) [reply src/test/projects.test.js:73]: Done

File: src/test/projects.test.js (hunk 24 above)
Comment (SteampunkMatu) [reply src/test/projects.test.js:80]: Done

File: src/test/projects.test.js (hunk 25 above)
Comment (SteampunkMatu) [reply src/test/projects.test.js:87]: Done

File: src/test/projects.test.js (hunk 26 above)
Comment (SteampunkMatu) [reply src/test/projects.test.js:94]: Done

File: src/test/projects.test.js (hunk 30 above)
Comment (SteampunkMatu) [reply src/test/projects.test.js:124]: Done

File: src/test/projects.test.js (hunk 31 above)
Comment (SteampunkMatu) [reply src/test/projects.test.js:127]: Done

File: src/test/projects.test.js (hunk 29 above)
Comment (SteampunkMatu) [reply src/test/projects.test.js:136]: Done

File: src/test/projects.test.js (hunk 28 above)
Comment (SteampunkMatu) [reply src/test/projects.test.js:123]: Done

File: src/test/projects.test.js (hunk 27 above)
Comment (SteampunkMatu) [reply src/test/projects.test.js:108]: Done

File: src/seed/projects.js [hunk 32]
```diff
@@ -0,0 +1,62 @@
...
+export default [{
+  _id: mongoose.Types.ObjectId('6352dae0282ae0599b7acd92'),
+  name: 'Coca Cola Webpage',
+  clientName: 'Coca Cola',
+  description: 'New web page for coca cola',
+  startDate: '2022-03-22T03:00:00.000Z',
+  endDate: '2022-03-25T03:00:00.000Z',
+  employees: [
+    {
+      _id: mongoose.Types.ObjectId('634c68f3658f142935ea7f6e'),
```
[the remark this answers is not in the document]
Comment (SteampunkMatu) [reply src/seed/projects.js]: Done

[the remark this answers is not in the document]
Comment (lmalvasia) [reply src/seed/projects.js]: @DamianPalavecino good catch 👏

**Review by lmalvasia — APPROVED** [review]

**Review by lmalvasia — APPROVED** [review]

**Pull request merged**